### PR TITLE
Fused weightnorm for ATen

### PR DESCRIPTION
**Pull request by mcarilli**
This PR contains a C++ implementation of weight norm.  The user-side exposure of weight norm through torch.nn.utils.weight_norm is unchanged.

If running on the GPU, and the norm is requested over the first or last dimension of the weight tensor, the forward pass is carried out using the fused kernels I wrote for our Fairseq GTC hero run, which offer superior performance to primitive ops and superior numerical stability when running in FP16.  In the common case that the backward pass is not itself constructing a graph (ie not attempting to set up double backward) the backward pass will be carried out using another fused kernel.  If the backward pass is constructing a graph, an alternate code path is taken, which does the math using differentiable primitive ops. In this way, the implementation allows double backward, even if the fused kernel was used in forward (although in this case, you don't benefit from the performance and stability of the fused backward kernel).

If running on the CPU, or if norming over an interior dim, the forward pass is carried out using double-differentiable primitive ops.

Figuring out how to generate all the right plumbing for this was tricky, but it was a fun experience learning how the autogenerator works and how the graph is constructed.  Thanks to @colesbury for useful guidance on this front.

I do have a few lingering questions:

- Should I unify my return statements (ie by default-constructing Tensors outside if blocks and using operator= within)?
- What is the significance of `non_blocking` when calling e.g. `auto norms = saved_norms.to(saved_g.type().scalarType(), non_blocking=True/False);`?  I am currently omitting `non_blocking`, so it defaults to False, but I didn't see any associated synchronizes on the timeline, so I'm wondering what it means.
- Is there an "official" mapping from at::ScalarTypes to corresponding accumulate types, as there are for the PODs + Half in [AccumulateType.h](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/AccumulateType.h)?  I looked for an equivalent mapping for ScalarTypes, didn't find one, and ended up rigging it myself (`  at::ScalarType AccType = g.type().scalarType() == at::ScalarType::Half ? at::ScalarType::Float : g.type().scalarType();`).
- Are sparse tensors a concern?  Should I include another check for sparse tensors in the `_weight_norm` entry point, and send those along the fallback CPU path as well?

### CODE REVIEW

File: aten/src/ATen/native/WeightNorm.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+Tensor norm_except_dim(const Tensor & v, int64_t pow, int64_t dim)
+{
+  // I assume tensor.contiguous(), view(), norm(), etc. here will dispatch through VariableType.
+  if(dim == -1)
+    return v.norm(pow);
+  else if(dim == 0)
+  {
+    std::vector<int64_t> output_size(v.dim(), 1);
+    output_size[0] = v.size(0);
+    return v.contiguous().view({v.size(0), -1}).norm(pow, 1).view(output_size);
```
Comment (SsnL) [inline aten/src/ATen/native/WeightNorm.cpp]: suggestion: use reshape here instead of contiguous + view.

Comment (mcarilli) [reply aten/src/ATen/native/WeightNorm.cpp]: Thank you for the quick review!  I had a couple reasons for choosing this pattern:
1.  I'm mimicking the original Python:  https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/weight_norm.py#L13
2.  norm_except_dim is accessible from the Python, and my implement [uses it](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/weight_norm.py#L13)) on the Python side, not just in _weight_norm on the C++ side.  IMO it's also a useful independent Python function to have.  For these two reasons, in general it's possible the incoming v may be noncontiguous, have pathologically permuted indices, etc.  I need the tensor's raw data to be C-style contiguous, so that A. it can be viewed with all but one dimension flattened and B.  the norm will be performant.  I'm not sure if reshape does all these things, so I decided to stick with what I knew would work.
3.  The ops in here need to be (double)differentiable.  I'm also not sure how reshape interacts with autograd (I expect it's fine and i'm being paranoid).

If 2. and 3. are not actually problems with reshape, I can use reshape instead.

Comment (mcarilli) [reply aten/src/ATen/native/WeightNorm.cpp]: The subsequent norm op involves (at least) 3 tensor-size round trips to global memory for a non-persistent implementation (2 reads and a write), so it's worth making v C-style contiguous for coalesced memory accesses during the norm.  Unfortunately reshape() is not strict enough to guarantee C-style contiguity in all cases:
```
>>> a = torch.Tensor([[1,2],[3,4]])
>>> a.t().reshape(a.size(0), -1).stride()
(1, 2)
>>> a.t().contiguous().view(a.size(0), -1).stride()
(2, 1)
```

**Comment by zou3519** [comment]
Native files shouldn't need to be registered. Have you tried a combination of:
- `python setup.py clean`
- rebuild
- `git pull --rebase upstream master` (pull in changes from master)

**Comment by houseroad** [comment]
@mcarilli I think https://github.com/mcarilli/pytorch/pull/1 should fix your onnx problem :-)

**Comment by bddppq** [comment]
@mcarilli The ROCM hcc compiler has some difficulties on doing template type deductions, so you need to annotate the type params of the templated kernel function like here https://github.com/pytorch/pytorch/blob/f9595e756e7dc64d9adb7951a31f5fdc9302cf99/aten/src/ATen/native/cuda/Dropout.cu#L121

@Jorghi12 @iotamudelta

**Comment by mcarilli** [comment]
@houseroad @bddppq Thank you very much!  I'll work on getting these changes integrated and hopefully resubmit today.

**Comment by houseroad** [comment]
@mcarilli no problem, you directly merge my pr to this branch :-)

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -1809,6 +1809,27 @@
...
 
+- func: norm_except_dim(Tensor v, int64_t pow=2, int64_t dim=0) -> Tensor
+  variants: function
+
+# VariableType::_weight_norm does not want to be given a gap in the autograd graph,
+# so we don't define "dispatch" variants for it.
+- func: _weight_norm(Tensor v, Tensor g, int64_t dim=0) -> Tensor
+  variants: function
+
+- func: weight_norm_cuda_interface(Tensor v, Tensor g, int64_t dim=0) -> (Tensor, Tensor)
```
[the remark this answers is not in the document]
Comment (mcarilli) [reply aten/src/ATen/native/native_functions.yaml]: I added the underscore decorations, but the latest push doesn't appear to be retriggering the CI (possibly because the PR is already approved).  For what it's worth I did build and run test_autograd.py one last time before pushing.

Thanks again for all your help!

**Comment by ezyang** [comment]
Don't worry about the CircleCI results.

**Comment by weiyangfb** [comment]
is this ready to merge? @ezyang
